File: src/ca/mindmagic/game/map/grid/GridMap.java
package ca.mindmagic.game.map.grid;

import ca.mindmagic.game.map.grid.Coordinate;
import ca.mindmagic.game.map.grid.Grid;
import java.util.Set;
import java.util.stream.Collectors;

import javafx.geometry.Point2D;

public abstract class GridMap {

  protected int height, width;
  protected Grid grid;
  protected boolean wrapVertical, wrapHorizontal;

  public GridMap(Grid grid, int height, int width, boolean wrapVertical, boolean wrapHorizontal) {
    this.grid = grid;
    this.height = height;
    this.width = width;
    this.wrapVertical = wrapVertical;
    this.wrapHorizontal = wrapHorizontal;
  }

  public Double[] verticesOf(Coordinate c) {
    return verticesOf(c.getRow(), c.getCol());
  }

  public abstract Double[] verticesOf(int row, int col);

  public abstract Point2D centerOf(int row, int col);

  public Point2D centerOf(Coordinate location) {
    return centerOf(location.getRow(), location.getCol());
  }

  public Set<Coordinate> neighborsOf(Coordinate location) {
    return neighborsOf(location.getRow(), location.getCol());
  }

  public abstract Set<Coordinate> neighborsOf(int row, int col);
  public abstract int range(Coordinate c1, Coordinate c2);

  @Deprecated
  public abstract Set<Coordinate> getLocations();

  public int getHeight(){
    return height;
  }

  public int getWidth(){
    return width;
  }

  public boolean locationExistsOnMap(Coordinate location){
    return locationExistsOnMap(location.getRow(), location.getCol());
  }

  public Set<Coordinate> getArea(Coordinate c, int radius){
    return getArea(c.getRow(), c.getCol(), radius);
  }

  public Set<Coordinate> getArea(int row, int col, int radius) {
    Set<Coordinate> area = grid.getArea(mapToGridCoordinate(row, col), radius).stream()
        .map(this::gridToMapCoordinate)
        .collect(Collectors.toSet());
    if (wrapHorizontal) {
      area = area.stream().map(this::wrapWidth).collect(Collectors.toSet());
    }
    if (wrapVertical) {
      area = area.stream().map(this::wrapHeight).collect(Collectors.toSet());
    }
    return area.stream().filter(this::locationExistsOnMap).collect(Collectors.toSet());
  }

  protected Coordinate mapToGridCoordinate(Coordinate c) {
    return mapToGridCoordinate(c.getRow(), c.getCol());
  }

  protected abstract Coordinate mapToGridCoordinate(int row, int col);

  protected Coordinate gridToMapCoordinate(Coordinate c) {
    return gridToMapCoordinate(c.getRow(), c.getCol());
  }

  protected abstract Coordinate gridToMapCoordinate(int xAxis, int yAxis);

  public boolean locationExistsOnMap(int row, int col){
    if (row < 0 || row > height - 1) return false;
    if (col < 0 || col > width - 1) return false;
    return true;
  }

  /**
   * The angle of a line between the center of the origin to the center of the target. This
   * can be useful for unit facing calculations
   *
   * @param origin the starting location
   * @param target the end point
   * @return the angle in degrees
   */
  public double angleOf(Coordinate origin, Coordinate target) {
    Point2D a = centerOf(origin);
    Point2D b = centerOf(target);
    return a.angle(centerOf(origin.getRow(), origin.getCol() + 1), b);
  }

  private Coordinate wrapHeight(Coordinate c) {
    int yOffset = c.getRow();
    if (yOffset < 0) {
      return new Coordinate(height + yOffset, c.getCol());
    }
    if (yOffset >= height) {
      return new Coordinate(yOffset - height, c.getCol());
    }
    return c;
  }

  private Coordinate wrapWidth(Coordinate c) {
    int xOffset = c.getCol();
    if (xOffset < 0) {
      return new Coordinate(c.getRow(), width + xOffset);
    }
    if (xOffset >= width) {
      return new Coordinate(c.getRow(), xOffset - width);
    }
    return c;
  }
}

File: src/ca/mindmagic/game/map/grid/GridMap.kt
package ca.mindmagic.game.map.grid

import ca.mindmagic.game.map.grid.pattern.Pattern
import java.util.stream.Collectors

open class GridMap @JvmOverloads constructor(pattern: Pattern,
                                             _width: Int, _height: Int,
                                             val wrapHorizontal: Boolean = false, val wrapVertical: Boolean = false) : Grid(pattern) {

    /*
        for wrapped maps, it is important to have even dimensions in order for edges to match up cleanly
        neighborsOf and range will possibly give undefined behavior if dimension is odd
     */
    // TODO consider forcing dimensions to be even (+ _height % 2)
    val height = if (_height < 1) 1 else _height
    val width = if (_width < 1) 1 else _width



    override fun neighborsOf(location: Coordinate): Set<Coordinate> {
        return pattern.neighborsOf(location).stream()
                .map { c -> wrapIfNeeded(c) }
                .filter { c -> contains(c) }
                .collect(Collectors.toSet())
    }

    fun contains(location: Coordinate): Boolean {
        return contains(location.row, location.col)
    }

    fun contains(row: Int, col: Int): Boolean {
        if (row < 0 || row >= height) return false
        if (col < 0 || col >= width) return false
        return true
    }

    /*
        It is important to note that wrapped maps need special consideration for range calculations as there is
        more than one possible path from source to target
     */
    override fun range(source: Coordinate, target: Coordinate): Int {
        // TODO optimize for use with wrapped edges. Should at least cap iterations of loop at max(width, height)
        return super.universalRange(source, target)  // quick fix but not optimized
    }

    private fun wrappedRange(source: Coordinate, target: Coordinate): Int {
        var range = range(source, target)
        if (wrapHorizontal) {

        }
        if (wrapVertical) {

        }
        return range
    }

    private fun wrapIfNeeded(c: Coordinate): Coordinate {
        var result = c
        if (wrapVertical) result = wrapVertical(result)
        if (wrapHorizontal) result = wrapHorizontal(result)
        return result
    }

    private fun wrapVertical(c: Coordinate): Coordinate {
        var yOffset = c.row
        while (yOffset < 0) {
            yOffset = height + yOffset       // it is important that height never be less than 1 or this will never exit
        }
        while (yOffset >= height) {
            yOffset -= height
        }
        return Coordinate(yOffset, c.col)
    }

    private fun wrapHorizontal(c: Coordinate): Coordinate {
        var xOffset = c.col
        while (xOffset < 0) {
            xOffset = width + xOffset       // it is important that width never be less than 1 or this will never exit
        }
        while (xOffset >= width) {
            xOffset -= width
        }
        return Coordinate(c.row, xOffset)
    }
}

